test/python/circuit/test_circuit_operations.py

```diff
@@ -550,6 +550,17 @@ def test_barrier(self):
 
         self.assertEqual(qc, expected)
 
+    def test_barrier_in_context(self):
+        """Test barrier statement in context, see gh-11345"""
+        qc = QuantumCircuit(2, 2)
+        qc.h(0)
+        with qc.if_test((qc.clbits[0], False)):
+            qc.h(0)
+            qc.barrier()
+
+        operation_names = [c.operation.name for c in qc]
+        assert "barrier" not in operation_names
```

Comment:
```suggestion
        self.assertNotIn("barrier", operation_names)
```
we don't use `pytest`, so `unittest` methods give us nicer error messages than bare `assert` (`pytest` does magic introspection to rewrite `assert`s).